SafetyFirst/ResultModule.fs

```diff
@@ -0,0 +1,467 @@
+  /// </summary>
+  [<CompiledName("$notForC#_createContext")>] // since the compiled name isn't valid from C#, it effectively makes it visible only from F#
+  let createContext errContext result = 
+    result
+    |> Result.mapError (fun e -> 
+      { Error = e; Context = [errContext] })
+
+  /// <summary>
+  /// Expands on the error context created by the <c>createContext</c> function.
+  /// </summary>
+  [<CompiledName("$notForC#_addContext")>]
+  let addContext newContext result = 
+    result
+    |> Result.mapError (fun ({ Context = context } as error) -> 
+      { error with Context = newContext :: context })
+
+  /// <summary>
+  /// If all the Results are ok, "unwraps" the ok values and passes them
+  /// to the function given, returning the result of that function.  If any Result 
+  /// is error, returns the first error without calling the function given. 
+  /// </summary>
+  [<CompiledName("$notForC#_bind2")>]
+  let bind2 onOk result1 result2 =
+    result {
+      let! r1 = result1
+      let! r2 = result2
+      return! onOk r1 r2
+    }
+
+  /// <summary>
+  /// If all the Results are ok, "unwraps" the ok values and passes them
+  /// to the function given, returning the result of that function.  If any Result 
+  /// is error, returns the first error without calling the function given. 
+  /// </summary>
+  [<CompiledName("$notForC#_bind3")>]
+  let bind3 onOk result1 result2 result3 =
+    result {
+      let! r1 = result1
+      let! r2 = result2
+      let! r3 = result3
+      return! onOk r1 r2 r3
+    }
+
+  /// <summary>
+  /// If all the Results are ok, "unwraps" the ok values and passes them
+  /// to the function given, returning the result of that function.  If any Result 
+  /// is error, returns the first error without calling the function given. 
+  /// </summary>
+  [<CompiledName("$notForC#_bind4")>]
+  let bind4 onOk result1 result2 result3 result4 =
+    result {
+      let! r1 = result1
+      let! r2 = result2
+      let! r3 = result3
+      let! r4 = result4
+      return! onOk r1 r2 r3 r4
+    }
+
+  /// <summary>
+  /// Collects a sequence of Results into a single Result of the sequence of values.
+  /// If all of the Results are Ok, returns an Ok of the sequence of contained values.  
+  /// If any of the Results are Error, returns an Error of the sequence of contained 
+  /// Errors (and throws away any Ok values). 
+  /// <c>collect [Ok 1; Ok 2; Ok 3]</c> would return <c>Ok [1; 2; 3]</c>, but 
+  /// <c>collect [Ok 1; Error "err"; Error "fail"]</c> would return <c>Error ["err"; "fail"]</c>
+  /// </summary>
+  [<CompiledName("$notForC#_collect")>]
+  let collect results =
+    Seq.fold 
+      (fun state element -> 
+        match state, element with
+        | (Ok xs, Ok x) -> Ok (Seq.append xs [x])
+        | (Ok _, Error x) -> Error (seq [x]) 
+        | (Error xs, Error x) -> Error (Seq.append xs [x])
+        | (Error xs, Ok _) -> Error xs)
+      (Ok (seq []))
+      results
```

Comment:
I'd like to switch this implementation to use Lists (like we did for the Option version).  It'll perform much faster and is guaranteed finite.

Reply:
✔️